1단계 코드 크루  리뷰

BadaHertz52 opened the pull request:
## 배포
[🌟 배포 페이지 바로가기](https://badahertz52.github.io/react-shopping-products/)

## 미션 구현 계획 시 사용한 피그잼
[🔗바로가기](https://www.figma.com/board/gBgHYkUPB0ibIJydDePZiR/Untitled?node-id=0-1)

## 피드백 받고 싶은 부분
### 장바구니 목록 상태 관리와 props drilling
[🔗 관련 코멘트 바로가기](https://github.com/BadaHertz52/react-shopping-products/pull/1/files#r1621585509)

### 장바구니 api 요청 오류 처리
[🔗관련 코멘트 바로가기](https://github.com/woowacourse/react-shopping-products/pull/9/files?file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx&show-viewed-files=true#r1620236309)

### 장바구니 빼기 테스트
[🔗 관련 코멘트 바로가기](https://github.com/BadaHertz52/react-shopping-products/pull/1/files#r1621586255)

BadaHertz52 committed:
ci: prettier, eslint 설정 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
chore : 라이브 쉐어를 위한 package.json 설정 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
docs: 기능 구현 목록 작성

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 글로벌 스타일, reset 스타일 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
chore: 절대 경로 설정

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
fix: global 스타일 확장자 변경

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: Header 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: Layout 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
chore: 절대 경로 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: Dropdown 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: CartActionButton 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: Header zIndex 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: CartActionButton 퍼블리싱

- Icon svg 파일 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: ProductCard 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: ProductList 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: PageListPage 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: scr-only 스타일링

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
test : 상품목록 mock data 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor: html lang 수정

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: App에 Layout ,ProductListPage 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor : main.tsx에 prettier 적용

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
chore: 스크롤 커스텀 스타일

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: fallback -CartAction 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: CartBadgeButton의 수량관련 퍼블리싱

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: Header에 CartAction fallback 적용

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: ProductList 스크롤 적용

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: Layout padding bottom 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

jinhokim98 committed:
feat: desktop width값 변경

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
refactor: product list margin bottom 추가

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
chore: env 관련 셋팅

Co-authored-by: badahertz52 <badahertz52@gmail.com>

BadaHertz52 committed:
feat: api 관련 유틸함수, 상수 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor: 필터링 타입명 변경 및 select에 name 추가

- 타입명 변경 : Sort->ProductSort

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 상품 목록 불러오는 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 상품 목록 핉터링 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat : 불러온 상품 목록 데이터 렌더링

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 상품 목록 api. 요청 옵션에 page 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 무한 스크롤 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 필터링 옵션 변경 시, 스크롤 위치 상단으로 변경하는 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: api 로딩과 errorBoundary 기능 추가

 - react-error-boundary  패키지 설치
- api 로딩과 errorBoundary 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
fix: endPoint 수정

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

jinhokim98 committed:
refactor: loading, error 로직 useFetch로 분리

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
feat: 무한스크롤 시 4개 씩 불러오는 기능 구현

Co-authored-by: badahertz52 <badahertz52@gmail.com>

BadaHertz52 committed:
refactor: fetch 유틸함수 수정

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 장바구니 담기 api 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor: fetchGetCartItems 수정

- 네이밍 변경
- 파라미터 추가
- 반환값 변경

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 장바구니 담기, 빼기 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

jinhokim98 committed:
test: msw 초기셋팅 작업

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
refactor: stackProducts 훅으로 분리

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
refactor: handler.ts 폴더 내로 이동

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
refactor: useStackProducts 훅 안으로 isLast, productLength를 넘겨줌으로써 마지막 페이지일 때 fetch 되지 않도록 변경

Co-authored-by: badahertz52 <badahertz52@gmail.com>

jinhokim98 committed:
test: useStackProducts 테스트

Co-authored-by: badahertz52 <badahertz52@gmail.com>

BadaHertz52 committed:
refactor: filterProducts 훅으로 분리

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
test : useFilterProducts 테스트

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor: useCartAction 훅 생성

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: cartItems에 대한  context 관련 기능 추가

- CartItemsContext , useTargetContext 생성
- cartItems에 대한  context 관련 기능 적용

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor: 훅 테스트 파일 위치 변경

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
refactor: serverResponse 유틸함수 분리

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
test: 장바구니 담기/빼기 테스트

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 모달 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 장바구니 api 오류 시 토스트 모달 띄우는 기능 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
feat: 페이지 렌더링 시 api 오류에 대한 error fallback 추가

Co-Authored-By: JinHo Kim <81083461+jinhokim98@users.noreply.github.com>

BadaHertz52 committed:
fix: vitest.config.ts 오류 수정

BadaHertz52 committed:
fix: css module global 오류 수정

BadaHertz52 committed:
chore: 빌드를 위한 추가 설정

BadaHertz52 committed:
docs: 리드미 수정 및 필요한 사진 추가

BadaHertz52 commented on src/hooks/useCartAction.ts:
```diff
@@ -0,0 +1,70 @@
+  useEffect(() => {
+    setCartActionError(getFetchResult.error || postFetchResult.error || deleteFetchResult.error);
+  }, [getFetchResult.error, postFetchResult.error, deleteFetchResult.error]);
+
```
장바구니 목록 불러오기와 담기/빼기에 대한 api 요청 오류 시 toast modal을 띄워주기 위해 error를 반환해주고 있어요.

일단 돌아가기 위해서 구현을 했는데. 코드가 깔끔하지 않아요.

해당 코드 리팩토링을 어떻게 하면 좋을까요?

soi-ha replied:
코드를 보고 개선 방향을 잡아드릴 수는 없지만, 저희의 경우에는 errorToast를 띄우는 것도 context를 활용했어요!
상위 요소에 context가 존재하니까 하위 요소의 api를 요청하는 부분에서 error가 발생하면 showError라는 함수( 에러 토스트를 보여주게 하는)가 실행되도록 했습니다.

BadaHertz52 commented on src/App.tsx:
```diff
@@ -1,11 +1,56 @@
+      <Header cartItemsLength={cartItems.length} headerRef={headerRef} />
+      <Layout>
+        <ErrorBoundary FallbackComponent={({ error }) => <PageRequest error={error} />}>
+          <CartItemsContext.Provider value={{ handleCartAction }}>
```
상품의 장바구니 담기/빼기 클릭 시 api요청 후 결과에 맞추어  Header 속 장바구니 아이콘에 장바구니 목록 수량이 업데이트 되어야해서 

장바구니 담기/빼기 + 장바구니 목록 상태를 업데이트해주는 handleCartAction함수를 context api로 내려주고 있어요.

크루는 해당 기능을 어떻게 구현하셨나요?

soi-ha replied:
저희도 장바구니 api인 cart-tiems의 데이터는 context로 관리하고 있습니다!

BadaHertz52 commented on src/hooks/tests/useCartAction.test.ts:
```diff
@@ -0,0 +1,57 @@
+          cartItem: undefined,
+        });
+        // TODO: 같은 get요청에 대해 테스트 케이스에 따라 다른 mock data 넣는 법....
+        //expect(result.current.cartItems.length).toBe(0);
```
장바구니 목록을 get하는 handler는 하나인데, 담기와 빼기 시 사용해야하는 목 데이터를 다르게 설정해야하는데 방법을 찾지 못했어요.

크루는 delete 테스트를 어떻게 진행하셨나요?

soi-ha replied:
와우.. 저희는 담기와 빼기에 대한 테스트는 진해아지 않아서..^^.. 드릴 말씀이 없숩니다..

soi-ha commented on src/hooks/tests/useFilteredProducts.test.ts:
```diff
@@ -0,0 +1,88 @@
+import useFetch from '../useFetch';
+import useFilteredProducts from '../useFilteredProducts';
+
+describe('상품 목록 필터링 테스트', () => {
```
와... handler에 필터링 해두셨네요..? 이렇게 테스트 하셨군요..? 저희는 깔끔하게 안했는데.. 최고..👍

soi-ha commented on src/hooks/useStackProducts.ts:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * 무한 스크롤 시 상품 목록을 추가해서 넣어주는 기능
+   */
+  const getStackedProducts = async (page: number) => {
```
깔-끔해서 보기 편하고 너무 좋네요..👍

soi-ha commented on src/hooks/useToastModalAnimation.ts:
```diff
@@ -0,0 +1,36 @@
+import { useLayoutEffect, useState } from 'react';
+
+export default function useToastModalAnimation({ isOpen, closeModal }: { isOpen: boolean; closeModal: () => void }) {
```
진짜 이렇게 까지 분리를 할 수 있을지 처음 알았어요.
너무 깔끔하고 좋네요!

soi-ha commented on src/pages/ProductListPage/ProductList/index.tsx:
```diff
@@ -0,0 +1,38 @@
+  const productListRef = useRef<HTMLElement | null>(null);
+
+  useEffect(() => {
+    if (products.length <= 20) {
```
products.length를 왜 **20개** 미만이 조건이 발생했는지 모르는 사람도 한 눈에 알아 볼 수 있도록 20을 상수화하는 것도 좋을 것 같아요!

soi-ha commented on src/apis/products.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  searchParams.append('sort', filtering.sort);
+  searchParams.append('page', String(page));
```
page의 타입이 number여서 string으로 만들어 주신 것 같은데, 제가 알기로는 그냥 숫자를 넣어도 된다고 알고 있습니다..! 혹시 string으로 변환하여 넣은 것에 대해 의미가 있는 걸 까요??

(근데 확실하진 않아욧.. 틀리먼 스루 해주십쇼..)

https://developer.mozilla.org/ko/docs/Web/API/URLSearchParams/get

soi-ha commented on src/appTypes/product.ts:
```diff
@@ -0,0 +1,15 @@
+export type Category = '' | 'fashion' | 'beverage' | 'electronics' | 'kitchen' | 'fitness' | 'books';
```
빈 문자열이 **전체** 때문에 있는 것으로 보이는데,
Category를 사용하는 곳에서 `Category | undefined`와 같은 형태가 아닌 빈 문자열로 넣어준 이유가 있을까요??

BadaHertz52 replied:
api 요청 보내는 것과 타입을 맞추었어요.

지금보니, ''를 ALL이라는 상수를 받드는 것이 선언적으로 더 좋을 것 같네요.

soi-ha commented on src/appTypes/response.ts:
```diff
@@ -0,0 +1,32 @@
+ * 서버 응답의 공통 구조를 정의하는 인터페이스
+ * @template T - response의 content타입
+ */
+export interface ServerResponse<T> {
```
오.. 이렇게 따로 분리하는 것도 좋네요~!

soi-ha commented on src/components/Dropdown/index.tsx:
```diff
@@ -0,0 +1,30 @@
+
+  return (
+    <>
+      <label className="scr-only" htmlFor={selectId}>
+        {label}
+      </label>
```
하루가 너무 좋아 할 것 같은!

BadaHertz52 replied:
이번 리뷰어는 유조....

soi-ha commented on src/components/Header/CartBadgeButton/style.module.css:
```diff
@@ -0,0 +1,36 @@
+
+.quantity.over::after {
+  vertical-align: super;
+  content: '\207A';
```
content 의 '/207A'는 뭐에용?

BadaHertz52 replied:
오른 쪽 위로 올라가는 작은 + 에요!!

soi-ha commented on src/components/Modal/ModalPortal/index.tsx:
```diff
@@ -0,0 +1,40 @@
+  children: ReactNode;
+}
+
+export default function ModalPortal(props: ModalPortalProps) {
```
Portal에 대해서 테코톡을 진행했던 바다라 그런지.. 포탈을 사용했군요..!
그러네요, 이번에 포탈을 errorToast를 띄울 때, 포탈을 사용하면 정말 유용하게 사용할 수 있었겠네요!

BadaHertz52 replied:
패키지 미션 복...부...ㅌ

soi-ha reviewed (commented):
홀홀홀... 
사실 개선 방향에 대해서 남기기에는 아직 코드를 읽고 보는 눈이 많이 모자라서.. 
대부분의 내용이 감탄과 깨달음이네요..!
도움이 되지 못한 것 같아서 미안해요 바다~! 🌊

그래도 코드 진짜 깔끔해서 보기 너무 좋았어요. 최고👍

BadaHertz52 committed:
chore : eslint 불필요한 설정 삭제

BadaHertz52 committed:
refactor: Fallbacks 컴포넌트명 수정

BadaHertz52 committed:
refactor: 절대 경로 설정 리팩토링

BadaHertz52 committed:
refactor: 토스트 모달에 대한 hook을 ToastModal 로 이동

-

BadaHertz52 committed:
feat: 토스트 모달의 위치에 대한 훅 생성

BadaHertz52 committed:
feat: 장바구니 api 오류에 대한 모달 컴포넌트 생성

CartActionErrorModal 생성

BadaHertz52 committed:
refactor: 장바구니 api 요청에 대한 훅 분리

-UseCartAction : 장바구니 담기/빼기 담당
-UseCartItemIds: 장바구니 목록 데이터 요청, 데이터를 이용해 key:product.id 이고 value는 cartItem.id인 Map 객체 상태 관리

BadaHertz52 committed:
refactor: 장바구니에 대한 api 요청 처리 리팩토링

- createPortal을 사용해 , 장바구니 api 요청이 이루어지는 ProductCard, Header에서 api 오류에 따라 CartActionErrorModal이 열리도록 함
- 장바구니 목록 자체를 props로 받는 게 아니라, context를 사용해  key:product.id 이고 value는 cartItem.id인 Map형태의 cartItemIds를 내려보내줌

BadaHertz52 committed:
feat: 상품 목록 페이지에 대한 스켈레톤 생성 및 적용

- 스켈레톤과  스켈레톤이 아닌 컴포넌트에서 동일하게 사용하는  상품 목록 스타일 ProductList.css에서 관리

BadaHertz52 committed:
feat: 무한 스크롤에 스켈레톤 적용

BadaHertz52 committed:
refactor: 불러오는 상품 목록 상수 처리

BadaHertz52 committed:
test: useCartAction 훅 변경에 따른 테스트 코드 수정

BadaHertz52 committed:
refactor: fetchGetCartItems 역할 분리

-  fetchGetCartItems 는 api에서 데이터만 불러오는 역할로 책임을 축소함
- page =0에서 모든 장바구니 목록을 불러오는 지 계산은 useCartItemIds 에서 진행하도록 함

BadaHertz52 committed:
refactor: 불러오는 상품 목록 상수 추가 처리

BadaHertz52 committed:
refactor: product-list-wrapper의 max-height에 필요한 높이 변수화

BadaHertz52 committed:
refactor: 상품 목록 페이지 상수화

BadaHertz52 committed:
refactor: useFilteredProducts, useStackProducts 리팩토링

 - 각 훅에서 반환하는 함수에서 필요한 값을 파라미터로 받도록 수정

BadaHertz52 committed:
test :useFilteredProducts, useStackProducts 리팩토링에 따른 테스트 코드 수정

BadaHertz52 committed:
refactor: IntersectionObserver 관련 리팩토링

- UseIntersectionObserver 훅 생성
- IntersectionObserverArea 컴포넌트는 UseIntersectionObserver를 통해 옵져버를 실행하는 컴포넌트로 역할 분리

BadaHertz52 committed:
refactor: 프로덕션 코드 리팩토링

- 타입명 변경 :  ServerResponse -> ApiResponse
- 함수 선언 방식 변경: function -> const

BadaHertz52 committed:
refactor: 모달 zindex 변수 관리

BadaHertz52 committed:
refactor: 변경된 타입명 적용

- ServerResponse -> ApiResponse  변경 후, 타입명 변경 안된 부분 수정

BadaHertz52 committed:
fix: 상품 목록 20개 렌더링 시, 무한 스크롤용 스켈레톤 보이는 오류 해결

BadaHertz52 committed:
refactor: package.json hompage 위치 변경
